### PR TITLE
describes RBACGenerator

### LINKED ISSUE #875: [Proposal] User management part II

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**: this pull bring the part II of the user management. Describes `RBACGenerator`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #875 

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: docs/proposals/user-management.md [hunk 1]
```diff
@@ -11,9 +11,10 @@ User management describes how to manage user access to various resources like de
 ## Goals
 1. Describe how `kubermatic-server` is going to drive authorization decisions. 
 2. Utilise kubernetes `RBAC` mechanism as much as possible. 
+3. Describe how `RBAC` roles are generated and attached to project’s resources.
 
 ## Non-Goal
-1. Describe how `RBAC` roles are generated and attached to project’s resources. 
+1. Describes how user management will work inside consumer clusters.
```
Comment (p0lyn0mial) [inline docs/proposals/user-management.md:17]: this is quite important, every time you will download `kubeconfig` provided by the dashboard you will get `admin` access to cluster.

File: docs/proposals/user-management.md [hunk 2]
```diff
@@ -83,20 +84,30 @@ One thing worth mentioning is that the names of the `Groups`, `Roles` and `RoleB
...
+To hide complexity from developers and to present a more hierarchical view we will create `ProjectProvider` which will accept a modified version of `ClusterProvider`, the new component will also rely on `RBACGenerator`.  The new component will not only make sure that queries are executed in the context of a user but it will also encapsulate resources and will provide convenient methods to operate on them. In the event of a resource creation or deletion the `ProjectProvider` will rely on `RBACGenerator` to create necessary `RBAC` roles.
 
 ## RBACGenerator
-TBD
+`RBACGenerator` is a separate component responsible for creating `Role` and `RoleBindings` for every `Group` we wish to have. It can be implemented as an active
+reconciliation loop that constantly watches for project's resources. The controller pattern guarantees that required `RBAC` are in place
+even in the event of failure, otherwise enforcing I want to "create a resource AND attach `RBAC`" semantic could be hard.
+
+Initially the component could maintain the hardcoded list of `Groups`. By convention the names of groups are `admin-projectIdentity`, `editor-projectIdentity` etc.
```
Comment (p0lyn0mial) [inline docs/proposals/user-management.md:105]: If this is not sufficient let me know I can put more details.

File: docs/proposals/user-management.md [hunk 3]
```diff
@@ -83,20 +84,30 @@ One thing worth mentioning is that the names of the `Groups`, `Roles` and `RoleB
...
 
 To recap when a user logs-in to our system this is what happens:
 - a user is authenticated
 - we retrieve a list of projects the user belongs to from `apiv1.User`
-- we query the system (using a build-in admin account) to find a matching instance of `ProjectGroups`
+- we query the system (using a build-in admin account) to find a matching instance of `ProjectGroupMappings`
```
Comment (p0lyn0mial) [inline docs/proposals/user-management.md]: I'm not sure if we need `ProjectGroupMappings` we could store this info in the `apiv1.User` resource.

File: docs/proposals/user-management.md [hunk 4]
```diff
@@ -83,20 +84,30 @@ One thing worth mentioning is that the names of the `Groups`, `Roles` and `RoleB
...
+Initially the component could maintain the hardcoded list of `Groups`. By convention the names of groups are `admin-projectIdentity`, `editor-projectIdentity` etc.
+The names of `Role` and `RoleBindings` are `resourceKind-groupName-role-projectIdentity` and `resourceKind-groupName-binding-projectIdentity` respectively.
+When the controller detects a new resource it will generate a pair of `Role` and `RoleBindings`. In order to do this it needs to determine
+the list of valid `verbs` for each `Group`. This part could also be hardcoded and implemented as a `mapper` that accepts the `groupName`and spits out the `verbs`.
+The next step requires `apiGroup`, `resourceKind`, `resourceName` and `verbs` to generate valid `Role`. The last step is to create `RoleBindings` this part
+needs the name of the `Role` and the subject which in our case is the `groupName`.
+
 
 ## Types
 TBD
```
Comment (mrIncompetent) [inline docs/proposals/user-management.md:114]: Should we define them within this PR or in a follow up?
I would prefer follow up atm

File: docs/proposals/user-management.md (hunk 2 above)
Comment (mrIncompetent) [reply docs/proposals/user-management.md:105]: Fine for me as it is

File: docs/proposals/user-management.md (hunk 3 above)
Comment (mrIncompetent) [reply docs/proposals/user-management.md]: I would prefer avoiding the additional type

File: docs/proposals/user-management.md (hunk 4 above)
Comment (p0lyn0mial) [reply docs/proposals/user-management.md:114]: I'd prefer follow up.

File: docs/proposals/user-management.md (hunk 3 above)
Comment (p0lyn0mial) [reply docs/proposals/user-management.md]: would you agree to storing this info in the `apiv1.User` type ? Something like the following:
```
...
[]ListForEachProject{
 projectIdentity: value,
 group:value,
}
```

Comment (mrIncompetent) [reply docs/proposals/user-management.md]: Yes, that look good!

**Review by mrIncompetent — APPROVED** [review]

**Comment by p0lyn0mial** [comment]
yay !

**Pull request merged**